File: src/main/java/blog/classic/javadsl/Greeter2.java
package blog.classic.javadsl;

import akka.actor.AbstractActor;

public class Greeter2 extends AbstractActor {

  public static class Greet {
  }

  public static class WhoToGreet {
    public final String who;

    public WhoToGreet(String who) {
      this.who = who;
    }
  }


  @Override
  public Receive createReceive() {
    return onMessage("hello");
  }

  private Receive onMessage(String currentGreeting) {
    return receiveBuilder()
      .match(WhoToGreet.class, this::onWhoToGreet)
      .match(Greet.class, msg -> onGreet(currentGreeting))
      .build();
  }

  private void onWhoToGreet(WhoToGreet whoToGreet) {
    getContext().become(onMessage("hello, " + whoToGreet.who));
  }

  private void onGreet(String currentGreeting) {
    System.out.println(currentGreeting);
  }

}

File: src/main/scala/blog/classic/scaladsl/Greeter2.scala
package blog.classic.scaladsl

import scala.io.StdIn

import akka.actor.Actor
import akka.actor.ActorRef
import akka.actor.ActorSystem
import akka.actor.Props

object Greeter2 {

  case object Greet
  final case class WhoToGreet(who: String)
}

class Greeter2 extends Actor {
  import Greeter2._

  override def receive = onMessage(currentGreeting = "hello")

  def onMessage(currentGreeting: String): Receive = {
    case WhoToGreet(who) =>
      context.become(onMessage(currentGreeting = s"hello, $who"))
    case Greet =>
      println(currentGreeting)
  }
}

object HelloWorldApp2 {
  import Greeter2._
  def main(args: Array[String]): Unit = {
    val system = ActorSystem("HelloWorld")
    try {
      val greeter: ActorRef = system.actorOf(Props[Greeter2], "greeter")
      greeter ! WhoToGreet("World")
      greeter ! Greet

      println("Press ENTER to exit the system")
      StdIn.readLine()
    } finally {
      system.terminate()
    }
  }
}
